### PR TITLE
[core][2/2] Worker resubscribe when GCS failed 

### LINKED ISSUE #24642: [core] Resubscribe for GCS pubsub from GCS side
When GCS restarts, it'll lose all subscription information since everything is stored in memory. Then in the runtime, we need to tell GCS what's currently being subscribed.

The previous method:
- We'll have a thread in core worker/raylet to check whether the GCS restarted or not.
- If it restarted, we'll send resubscribe request to GCS.

However, this is not working in these cases:
- GCS restarts happen so fast so the checker in raylet/core worker missed them.
- GCS doesn't restart, but just being lag due to network issues then, the resubscribe is not necessary.

Actually, GCS knows when a resubscribe is needed: when it restarts. So the PR here is to send a resubscribe request from GCS -> Raylet and Raylet will do the resubscription.

There are two parts for this to work:

- [ ] raylet send resubscription
- [ ] raylet ask core worker to send resubscription

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?

A follow-up PR from this one: https://github.com/ray-project/ray/pull/24628

In the previous PR, it fixed the resubscribing issue for raylet. But there is also core worker which needs to do resubscribing.

There are two ways of doing resubscribe:
1. When the client-side detects any failure, it'll do resubscribing.
2. Server side will ask the client to do resubscribing.

1) is a cleaner and better solution. However, it's a little bit hard due to the following reasons:

- We are using long-polling, so for some extreme cases, we won't be able to detect the failure. For example, the client-side received the message, but before it sends another request, the server-side restarts, and the client will miss the opportunity of detecting the failure. This could happen if we have a standby GCS that starts very fast and somehow the client-side has a lot of traffic and runs very slow.
- The current gRPC framework doesn't give the user a way to handle failure which might need some refactoring on this one.

We can go with this way once we have gRPC streaming.

This PR is implementing 2) which includes three parts:
- raylet: (https://github.com/ray-project/ray/pull/24628)
- core worker: (this pr)
- python

Correctness: whenever when a worker started, it'll register to raylet immediately (sync call) before connecting to GCS. So, we just need to send all restart rpcs to registered workers and it should work because:
- if the worker just started and hasn't registered with the raylet: it's ok, because the worker hasn't connected with GCS yet, so no need to do resubscribing.
- if the worker has registered with the rayelt: it's covered by the code path here.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/raylet/worker.cc [hunk 1]
```diff
@@ -96,17 +96,41 @@ int Worker::AssignedPort() const { return assigned_port_; }
...
+}
+
 void Worker::Connect(int port) {
   RAY_CHECK(port > 0);
   port_ = port;
   rpc::Address addr;
   addr.set_ip_address(ip_address_);
   addr.set_port(port_);
   rpc_client_ = std::make_unique<rpc::CoreWorkerClient>(addr, client_call_manager_);
+  if (notify_gcs_restarted_) {
```
Comment (mwtian) [inline src/ray/raylet/worker.cc:120]: Maybe call the other overload of `Worker::Connect()` instead?

File: src/ray/raylet/worker.h [hunk 2]
```diff
@@ -281,6 +283,8 @@ class Worker : public WorkerInterface {
   std::shared_ptr<TaskResourceInstances> lifetime_allocated_instances_;
   /// RayTask being assigned to this worker.
   RayTask assigned_task_;
+  /// If true, a RPC need to be sent to notify the worker about GCS restarting.
+  bool notify_gcs_restarted_;
```
Comment (mwtian) [inline src/ray/raylet/worker.h:287]: Default to `false`?

File: src/ray/raylet/node_manager.cc [hunk 3]
```diff
@@ -1081,7 +1081,20 @@ void NodeManager::ResourceDeleted(const NodeID &node_id,
...
+  // When GCS restarts, it'll notify raylet to do some initialization work
+  // (resubscribing). Raylet will also notify all workers to do this job. Workers are
+  // registered to raylet first (blocking call) and then connect to GCS, so there is no
+  // race condition here.
   gcs_client_->AsyncResubscribe();
+  auto workers = worker_pool_.GetAllRegisteredWorkers(true);
+  for (auto worker : workers) {
+    worker->AsyncNotifyGCSRestart();
+  }
+
```
Comment (mwtian) [inline src/ray/raylet/node_manager.cc:1093]: minor: remove this empty line?

**Review by mwtian — APPROVED** [review]
Is there a plan for Python subscribers?

**Comment by iycheng** [comment]
> Is there a plan for Python subscribers?

There are logging, error, dashboard that is using pubsub and the error is not that critical.
So for python, I plan to do it with the client-side resubscribe in the python layer. I'll get a RFC PR for some feedback.

**Pull request merged**